lib/Clam/Properties/BndCheck.cc

```diff
@@ -0,0 +1,348 @@
+    auto BasePair = memory_check_utils::getBasePtr(Ptr);
+    if (Value *BasePtr = BasePair.getPointer()) {
+      if (BasePair.getInt() == 1) {
+        // errs() << "Skipped " << *BasePtr << " because it is
+        // dereferenceable!\n";
+        return false;
+      }
+      // We've checked BasePtr in the current BB.
+      if (!m_seen.insert(BasePtr).second) {
+        // errs() << "Skipped " << *BasePtr << " because already checked!\n";
+        return false;
+      }
+    }
+    return true;
+  }
+
+  // Add bound check assertions with the following pattern:
+  // assert(offset >= 0)
+  // assert(size > offset)
+  // assert(size >= offset + size_of(value))
+  // where offset indicates the offset of current pointer
+  // the size means the total size of allcation
+  // sizeof(value) gets the bit-width of the element
+  void addBndCheckAssertion(const var_t &reg, const var_t &size,
+                            const var_t &offset, basic_block_t &bb,
+                            Instruction &I, unsigned size_of) {
+    if (m_params.check_only_typed_regions &&
+        reg.get_type().is_unknown_region()) {
+      return;
+    }
+    // We ignore for now the option
+    // m_params.check_only_noncyclic_regions.  To consider it the
+    // easiest way is to extend CrabStoreRefOps with information
+    // about the HeapAbstraction Region such as whether
+    // isSequence(), IsCyclic(), etc.
+
+    bb.assertion(offset >= number_t(0), getDebugLoc(&I));
+    bb.assertion(size > offset, getDebugLoc(&I));
+    bb.assertion(size >= offset + size_of, getDebugLoc(&I));
+    m_addedChecks++;
+    return;
+  }
+
+  // -- dereference a ptr with bound checks
+  void doDerefBndCheck(Instruction &I, basic_block_t &bb, var_t &reg) {
+    Value *key;
+    bool isLoad = false;
+    std::string output = "store";
+    if (isa<LoadInst>(I)) {
+      key = static_cast<LoadInst &>(I).getPointerOperand();
+      isLoad = true;
+      output = "load";
+    } else if (isa<StoreInst>(I)) {
+      key = static_cast<StoreInst &>(I).getPointerOperand();
+    } else {
+      CLAM_WARNING("Unexpected type of instruction for deference check");
+      return;
+    }
+    auto it = m_ref_bnd_map.find(key);
+    if (it == m_ref_bnd_map.end()) {
+      CLAM_WARNING("Could not find " + output + " ptr from map " << *key);
+    } else {
+      CRAB_LOG("cfg-bnd-check",
+               llvm::errs()
+                   << output + ": check size and offset are in bounds: " << I
+                   << "\n");
+      var_t ptr_obj_sz = it->second.first;
+      var_t ptr_offset = it->second.second;
+      Type *elemTy = key->getType()->getPointerElementType();
+      const llvm::DataLayout *dl = getInsDatalayout(I);
+      unsigned size_of = dl->getTypeSizeInBits(elemTy) / 8;
+      addBndCheckAssertion(reg, ptr_obj_sz, ptr_offset, bb, I, size_of);
+      crab_lit_ref_t lhs = m_lfac.getLit(I);
+      if (isLoad && lhs->getVar().get_type().is_reference()) {
```

Comment:
This code shouldn't be here. This code should be moved to `visitAfterLoad`. 
There two completely different things we do when we have a Load instruction. Before we actually added the corresponding Crab `load_ref` we check that the access is in-bounds. Then, once the `load_ref` happens we update `m_ref_bnd_map`.